e2e/callback_test.go

```diff
@@ -0,0 +1,372 @@
+	require.Equal(suite.T(), chainAOwner, suite.bridgeA.String())
+	chainBOwner := queryGetOwnerOf(suite.T(), suite.chainB, suite.cw721B.String(), "3")
+	require.Equal(suite.T(), chainBOwner, suite.chainB.SenderAccount.GetAddress().String())
+
+	// We query the data we have on the tester contract
+	// This ensures that the callbacks are called after all the messages was completed
+	// and the transfer was successful
+	testerDataOwnerA := queryTesterSent(suite.T(), suite.chainA, suite.testerA.String())
+	require.Equal(suite.T(), testerDataOwnerA, suite.bridgeA.String())
+	testerNftContract := queryTesterReceivedNftContract(suite.T(), suite.chainB, suite.testerB.String())
+	require.Equal(suite.T(), testerNftContract, suite.cw721B.String())
+	testerDataOwnerB := queryTesterReceived(suite.T(), suite.chainB, suite.testerB.String())
+	require.Equal(suite.T(), testerDataOwnerB, suite.chainB.SenderAccount.GetAddress().String())
+}
+
+func (suite *CbTestSuite) TestTimeoutedTransfer() {
+	memo := callbackMemo(nftSentCb(), "", nftReceivedCb(), "")
+	sendIcsFromChainA(suite, suite.cw721A.String(), "2", memo, false)
+	suite.coordinator.IncrementTimeBy(time.Second * 2001)
+	suite.coordinator.UpdateTime()
+	suite.coordinator.TimeoutPendingPackets(suite.path)
+
+	// Query the owner of NFT on cw721
+	chainAOwner := queryGetOwnerOf(suite.T(), suite.chainA, suite.cw721A.String(), "2")
+	require.Equal(suite.T(), chainAOwner, suite.testerA.String())
+	err := queryGetOwnerOfErr(suite.T(), suite.chainB, suite.cw721B.String(), "2")
+	require.Error(suite.T(), err)
+
+	// callbacks should update sender contract of the failed transfer
+	// so we query the contract to see who is the new owner
+	// if the query is working and owner is correct, we can confirm the callback was called successfully
+	testerDataOwnerA := queryTesterSent(suite.T(), suite.chainA, suite.testerA.String())
+	require.Equal(suite.T(), testerDataOwnerA, suite.testerA.String())
+
+	// Querying the receving end, should fail because we did not receive the NFT
+	// so the callback should not have been called.
+	err = queryTesterReceivedErr(suite.T(), suite.chainB, suite.testerB.String())
+	require.Error(suite.T(), err)
+}
+
+func (suite *CbTestSuite) TestFailedCallbackTransfer() {
+	memo := callbackMemo(nftSentCb(), "", failedCb(), "")
+	sendIcsFromChainA(suite, suite.cw721A.String(), "2", memo, true)
+
+	// Query the owner of NFT on cw721
+	chainAOwner := queryGetOwnerOf(suite.T(), suite.chainA, suite.cw721A.String(), "2")
+	require.Equal(suite.T(), chainAOwner, suite.testerA.String())
+	err := queryGetOwnerOfErr(suite.T(), suite.chainB, suite.cw721B.String(), "2")
+	require.Error(suite.T(), err)
+
+	// callbacks should update sender contract of the failed transfer
+	// so we query the contract to see who is the new owner
+	// if the query is working and owner is correct, we can confirm the callback was called successfully
+	testerDataOwnerA := queryTesterSent(suite.T(), suite.chainA, suite.testerA.String())
```

Comment:
Not sure, I understand this - what does this query mean and what is data owner?

Reply:
`queryTesterSent` is a storage on our tester contract, to test if the ack was called or not, and what data it holds.
As data I use the owner of the NFT when the ack callback was called, also to make sure the callback is called AFTER the NFT was transferred successfully.

IF this query, holds the same data as the query we do directly on the cw721 contract, then we know its a successful test.